homeassistant/components/roon/config_flow.py

```diff
@@ -25,36 +26,75 @@
 class RoonHub:
     """Interact with roon during config flow."""
 
-    def __init__(self, host):
-        """Initialize."""
-        self._host = host
+    def __init__(self, hass):
+        """Initialise the RoonHub."""
+        self._hass = hass
 
-    async def authenticate(self, hass) -> bool:
-        """Test if we can authenticate with the host."""
+    async def discover(self):
+        """Try and discover roon servers."""
+
+        def get_discovered_servers(discovery):
+            servers = discovery.all()
+            discovery.stop()
+            return servers
+
+        discovery = RoonDiscovery(None)
+        servers = await self._hass.async_add_executor_job(
+            get_discovered_servers, discovery
+        )
+        _LOGGER.debug("Servers = %s", servers)
+        return servers
+
+    async def authenticate(self, host, servers):
+        """Authenticate with one or more roon servers."""
         token = None
+        core_id = None
         secs = 0
-        roonapi = RoonApi(ROON_APPINFO, None, self._host, blocking_init=False)
-        while secs < TIMEOUT:
-            token = roonapi.token
+        if host is None:
+            apis = [
+                RoonApi(ROON_APPINFO, None, server[0], server[1], blocking_init=False)
+                for server in servers
+            ]
+        else:
+            apis = [RoonApi(ROON_APPINFO, None, host, blocking_init=False)]
+
+        while secs <= TIMEOUT:
+            # Roon can discover multiple devices - not all of which are proper servers, so try and authenticate with them all.
+            # The user will only enable one - so look for a valid token
+            auth_api = [api for api in apis if api.token is not None]
+
             secs += AUTHENTICATE_TIMEOUT
-            if token:
+            if auth_api:
+                core_id = auth_api[0].core_id
+                token = auth_api[0].token
                 break
+
             await asyncio.sleep(AUTHENTICATE_TIMEOUT)
 
-        token = roonapi.token
-        roonapi.stop()
-        return token
+        for api in apis:
+            await self._hass.async_add_executor_job(api.stop)
```

Comment:
Please make a function that iterates the APIs and stops them. Then schedule that function once on the executor.